src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/tasks/BlobProcessorTask.java

```diff
@@ -108,20 +110,31 @@ private void processZipFile(CloudBlobContainer container, String zipFilename)
                 }
             }
 
-            Envelope envelope = envelopeProcessor.processEnvelope(metadataStream);
+            envelope = envelopeProcessor.processEnvelope(metadataStream);
+            isUploadFailure = true;
 
-            try {
-                documentProcessor.processPdfFiles(pdfFiles, envelope.getScannableItems());
-                envelopeProcessor.markAsUploaded(envelope, container.getName(), zipFilename);
-
-                cloudBlockBlob.delete();
-            } catch (Exception exception) {
-                // catching any exception originated from document processor
-                envelopeProcessor
-                    .markAsUploadFailed(exception.getMessage(), envelope, container.getName(), zipFilename);
-                // rethrowing
-                throw exception;
-            }
+            documentProcessor.processPdfFiles(pdfFiles, envelope.getScannableItems());
+            envelopeProcessor.markAsUploaded(envelope, container.getName(), zipFilename);
+
+            cloudBlockBlob.delete();
+        } catch (Exception exception) {
+            markAsFailed(isUploadFailure, container.getName(), zipFilename, exception.getMessage(), envelope);
+
+            throw exception;
         }
     }
+
+    private void markAsFailed(
+        boolean isUploadFailure,
+        String container,
+        String zipFileName,
+        String message,
+        Envelope envelope
+    ) {
+        EnvelopeProcessor.FailureMarker marker = isUploadFailure
+            ? envelopeProcessor::markAsUploadFailed
+            : envelopeProcessor::markAsGenericFailure;
+
+        marker.markAsFailure(message, envelope, container, zipFileName);
+    }
```

Comment:
imho it would be clearer just do:
```
if (isUploadFailure) {
  envelopeProcessor.markAsUploadFailure(...)
} else {
  envelopeProcessor.markAsGenericFailure(...)
}
```
where this function is called (and remove the functional interface).

Or use a `switch` statement if there are more failure types in the future.

Reply:
No it is not. You have more than clear extraction of functional interface of what it's doing/what it's responsible of. Generally we mark the failure hence the method call. Signature is exactly the same and duplicating that in simple if/else call is simply introducing too many characters to read to clarify the correctness.

The correct implementation is in place disregarding the switch case. Ideally we would like to have general container of failure report and get rid of code branching entirely. Again - best to leave for later once bigger picture in mind